src/click/core.py

```diff
@@ -2731,7 +2731,8 @@ def _write_opts(opts: t.Sequence[str]) -> str:
         show_default_is_str = isinstance(self.show_default, str)
 
         if show_default_is_str or (
-            default_value is not None and (self.show_default or ctx.show_default)
+            (default_value or isinstance(default_value, bool))
```

Comment:
This fails for a default int value of 0, and any other falsey values.

Reply:
Oh.. obviously... I repushed, please verify is it better now.